torchvision/csrc/cpu/image/readpng_cpu.cpp

```diff
@@ -72,30 +74,74 @@ torch::Tensor decodePNG(const torch::Tensor& data) {
     TORCH_CHECK(retval == 1, "Could read image metadata from content.")
   }
 
-  int channels;
-  switch (color_type) {
-    case PNG_COLOR_TYPE_RGB:
-      channels = 3;
-      break;
-    case PNG_COLOR_TYPE_RGB_ALPHA:
-      channels = 4;
-      break;
-    case PNG_COLOR_TYPE_GRAY:
-      channels = 1;
-      break;
-    case PNG_COLOR_TYPE_GRAY_ALPHA:
-      channels = 2;
-      break;
-    case PNG_COLOR_TYPE_PALETTE:
-      channels = 1;
-      break;
-    default:
-      png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
-      TORCH_CHECK(false, "Image color type is not supported.");
+  int current_channels = png_get_channels(png_ptr, info_ptr);
+
+  if (channels > 0 && channels != current_channels) {
+    // TODO: consider supporting PNG_INFO_tRNS
+    bool is_palette = (color_type & PNG_COLOR_MASK_PALETTE) != 0;
+    bool has_color = (color_type & PNG_COLOR_MASK_COLOR) != 0;
+    bool has_alpha = (color_type & PNG_COLOR_MASK_ALPHA) != 0;
+
+    switch (channels) {
+      case 1: // Gray or Palette
+        if (has_alpha) {
```

Comment:
following my comment just above, I wonder if we shouldn't let `channels=1` to always return a grayscale image.

What do you think?

Reply:
Marking this as resolved this to keep the discussion in one place.